### ISSUE TITLE
Rehearsal revealed: The limits and merits of revisiting samples in continual learning

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#139]
## Motivation

- Analyze the impact of rehearsal (replay) in terms of trajectory in parameter space loss landscape

## TL; DR

- Since rehearsal memory is a small subset of the training data, it differs in distribution with previous task, but has some overlaps in the parameter space (***hence still useful***)
- Learning task 2 after task 1, the model weight at first moves towards the low loss region of task 2 but deflects to settle for the intersection between low loss region of rehearsal memory and task 2 data. (which can deviate from low loss region of task 1 data)
- Based on these observations, adding simple heuristic (training on rehearsal memory only for few more epochs) leads to better performance (moving *inwards* in the parameter space)

## Previous Work

- Rehearsal-based continual learning methods

## Method

![image](https://user-images.githubusercontent.com/46308407/146366540-6e20721e-bfc8-4ff4-9aad-05e35a099d50.png)

- Key observation: the weights at first move for lower loss region of next task, but then *deflects* to settle in between (where loss for both task 2 and rehearsal memory can be minimized)
- However, this might not be optimal for the whole task 1 data, as rehearsal memory only represents a small subset.

## Experiments

![image](https://user-images.githubusercontent.com/46308407/146366778-f0654842-68a3-4469-b26f-2bee0ce2837f.png)

- Looking at the red contour lines, we see that the *low loss region* for the task 1 data and rehearsal memory differ, so rehearsal-based continual learner could show bad generalization performance.
- Still, there is *some* overlap in the low loss region in parameter space, backing up performance gains from memory replays.

![image](https://user-images.githubusercontent.com/46308407/146367019-26406c6a-29b5-4a28-a47b-1f5dc28efc8e.png)

- Experiments on Mini-ImageNet also shows similar implications. Loss for the rehearsal memory remains very low, but that for task 1 data rises (=overfitting on the memory)

![image](https://user-images.githubusercontent.com/46308407/146367137-46f7cb7e-cdb2-4301-89cf-700de49a5b49.png)

- Visualization on the gradient norm also demonstrates similar overfitting behavior.

![image](https://user-images.githubusercontent.com/46308407/146367191-d72d6807-d67f-44f1-87c7-6471838c5a80.png)

- Based on these observations, the authors devised a simple heuristic to improve generalization: train only on the rehearsal memory for additional epochs so that the model weights move *inward* towards the low loss intersection of the tasks.
- Quantitative evaluation suggests that this does have some positive impacts.

![image](https://user-images.githubusercontent.com/46308407/146367351-b5a649a0-6087-49ac-81e4-b88fc21513f0.png)

- Analysis on the SOTAs. They benefit from rehearsal memory, but are nevertheless prone to overfitting,

## Comments

- Quite simple observational study on the loss behavior during rehearsal-based continual learning.
- The gist of this work, in my opinion, lies in figure 2, where they visualize the parameter trajectory in terms of loss values.
- The insights the authors intend to deliver are somewhat obvious, but nicely confirmed with adequate experiments and visualizations.
